### DISCUSSION TITLE
Should `M-SPC` clear selected candidate preview?

### DISCUSSION THREAD

**Discussion by jdtsmith** [minad/corfu#525] [Ideas]
Often I scroll down a bit in the list, and then realize I'm going to need to search instead.  But the candidate preview is in the way.  If I haven't gone far I can go back up to the first candidate.  Or I can reset the input with `ESC-ESC-ESC`, but that's a lot to type.  

I wonder if we could overload `M-SPC` to call `corfu-reset`?  Right now it does nothing in this scenario (well, it flashes the popup).  And this is the key combo I mentally reach for when I think "OK, I need to sharpen my search".   If there is no user-input at all yet, probably omitting the actual separator is also the right thing to do.

**Comment by minad** [comment]
Yes, let's do that. Feel free to commit this right away if the change is simple enough.

> If there is no user-input at all yet, probably omitting the actual separator is also the right thing to do.

You mean, if M-SPC is pressed at the prompt with no user input? I prefer to keep the current behavior - this way the user can initiate an Orderless search right away.

> Or I can reset the input with ESC-ESC-ESC, but that's a lot to type.

Regarding `ESC-ESC-ESC`, I think this is one of the bad Emacs defaults. My suggestion is to bind `<escape>` to `keyboard-escape-quit`, but this makes sense on the GUI only of course.

**Comment by jdtsmith** [comment]
> You mean, if M-SPC is pressed at the prompt with no user input?

No user input, _and_ previewing a selected candidate.  I actually don't recall how we test the latter. `corfu--preview-current-p`?  

So M-SPC would:

- just reset if previewing without any user input. I'm clearing out the selection to begin a search.
- insert separator if _not_ previewing (like now).  I'm starting or refining a search.
- reset + insert separator if previewing with (hidden) user input.  I.e. I'm continuing to refine an ongoing search.

**Reply by minad** [reply]
> No user input, and previewing a selected candidate. I actually don't recall how we test the latter. corfu--preview-current-p?

I am not sure if such a mixed behavior is ideal. I would have to test. I think the simplest solution would be to call `corfu-reset` always and never insert the separator. This might be unexpected for `M-SPC` users though? Alternatively always jump back to the prompt in `corfu-insert-separator` and always insert.

**Reply by minad** [reply]
After some testing I find DWIM behavior better. What about this?

~~~ elisp
(defun corfu-insert-separator ()
  "Insert a separator character, inhibiting quit on completion boundary.
See `corfu-separator' for more details."
  (interactive)
  (if (not (corfu--preview-current-p))
      (insert corfu-separator)
    (corfu--goto -1)
    (unless (or (= (car completion-in-region--data) (point))
                (= (char-before) corfu-separator))
      (insert corfu-separator))))
~~~

**Reply by jdtsmith** [reply]
This latest version seems great to me; it DoesWIM ;).

**Reply by minad** [reply]
Okay, added https://github.com/minad/corfu/commit/ac505abdc0d20eba2d1222f8552e4422338c2843.

**Comment by jdtsmith** [comment]
BTW, do you know if eglot's CAPF's caching behavior changed?  Now it seems to respond to a space separator by completing ignoring the user-entered component before the space, which I don't think used to happen.  

I am not wrapping its CAPF in in `capf-buster` anymore for a long time (which must have been for a reason).  No worries if not; I know you're not an eglot user.

**Reply by minad** [reply]
No, I am not aware of any such change. I'd prefer if we don't touch this area for the time being if you don't mind.
